plutus-core/untyped-plutus-core/src/Language/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -0,0 +1,505 @@
+import           Data.DList                                       (DList)
+import qualified Data.DList                                       as DList
+import           Data.Hashable                                    (Hashable)
+import           Data.STRef
+import           Data.Text.Prettyprint.Doc
+
+{- Note [Scoping]
+   The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
+   prerequisite. The CEK machine correctly handles name shadowing.
+-}
+
+data ExBudgetCategory fun
+    = BConst
+    | BVar
+    | BLamAbs
+    | BApply
+    | BDelay
+    | BForce
+    | BError
+    | BBuiltin         -- Cost of evaluating a Builtin AST node
+    | BBuiltinApp fun  -- Cost of evaluating a fully applied builtin function
+    | BAST
+    deriving stock (Show, Eq, Ord, Generic)
+    deriving anyclass (NFData, Hashable)
+instance Show fun => Pretty (ExBudgetCategory fun) where
+    pretty = viaShow
+
+instance ExBudgetBuiltin fun (ExBudgetCategory fun) where
+    exBudgetBuiltin = BBuiltinApp
+
+type TermWithMem uni fun = Term Name uni fun ExMemory
+
+{- Note [Arities in VBuiltin]
+The VBuiltin value below contains two copies of the arity (list of
+TypeArg/TermArg pairs) for the relevant builtin.  The second of these
+is consumed as the builtin is instantiated and applied to arguments,
+to check that type and term arguments are interleaved correctly.  The
+first copy of the arity is left unaltered and only used by
+dischargeCekValue if we have to convert the frame back into a term
+(see mkBuiltinApplication).  An alternative would be to look up the
+full arity in mkBuiltinApplication, but that would require a lot of
+things to be monadic (including the PrettyBy instance for CekValue,
+which is a problem.)
+-}
+
+-- 'Values' for the modified CEK machine.
+data CekValue uni fun =
+    VCon ExMemory (Some (ValueOf uni))
+  | VDelay ExMemory (TermWithMem uni fun) (CekValEnv uni fun)
+  | VLamAbs ExMemory Name (TermWithMem uni fun) (CekValEnv uni fun)
+  | VBuiltin            -- A partial builtin application, accumulating arguments for eventual full application.
+      ExMemory
+      fun
+      Arity             -- Sorts of arguments to be provided (both types and terms): *don't change this*.
+      Arity             -- A copy of the arity used for checking applications/instantiatons: see Note [Arities in VBuiltin]
+      Int               -- The number of @force@s to apply to the builtin.
+                        -- We need it to construct a term if the machine is returning a stuck partial application.
+      [CekValue uni fun]    -- Arguments we've computed so far.
+    deriving (Show, Eq) -- Eq is just for tests.
+
+type CekValEnv uni fun = UniqueMap TermUnique (CekValue uni fun)
+
+newtype CekBudgetSpender st uni fun = CekBudgetSpender
+    { unCekBudgetSpender
+        :: forall term s. ToExMemory term
+        => ExBudgetCategory fun -> ExBudget -> CekCarryingM st term uni fun s ()
+    }
```

Comment:
This is where we use the handle pattern. Or, rather, the inlined [`dep-t`](http://hackage.haskell.org/package/dep-t) pattern. Note that `CekCarryingM` references `CekBudgetSpender` and vice versa, so if we wanted to abstract the budgeting logic out, we'd need a CEK-unspecific `BudgetSpender` that gets instantiated to a CEK-specific one. I decided we don't need this generality: I don't see any other evaluator implementing budgeting any time soon, so I think we don't care about decoupling budgeting from the CEK machine right now (but it's nice to know that this is perfectly possible). Especially given the fact that such generalizations may slowdown the existing budgeting machinery. So I also went ahead and removed a few indirections (e.g. we now only have `CekExTally` and no general `ExTally`).

Reply:
Ah, yeah, I haven't implemented the same thing for logging. This PR is already somewhat big, so I'll do that in a subsequent PR.

Reply:
>  Not necessarily in the text, but just as a comment in the PR?

It comes in two flavours. One of them is explained in details [here](https://jaspervdj.be/posts/2018-03-08-handle-pattern.html). The other one is when your `Handle` is not just an opaque data type, but an actual function. This is what I'm going in this PR. `dep-t` above is a link that goes to a fairly readable description of the pattern (and by "inlined" I wanted to say "a particular instance of the general pattern").